File: lib/ai/providers/google.ts
// Adaptateur Google Gemini — API officielle (https://ai.google.dev/gemini-api/docs/image-generation).
// Modèles servis : gemini-2.5-flash-image ("Nano Banana"),
// gemini-3-pro-image-preview ("Nano Banana Pro") — édition multi-images.
//
// Schéma : POST /v1beta/models/{model}:generateContent (header
// `x-goog-api-key`) ; l'image source + les références partent en parts
// inline_data ; la réponse rend des parts inlineData (base64), converties
// ici en data URIs (le stockage objet arrive au jalon DB).
// Une image par requête : la quantité est servie par requêtes parallèles.
//
// Clé : GOOGLE_API_KEY (https://aistudio.google.com/apikey) — serveur uniquement.
import type { ProviderAdapter } from "../types";
import { ProviderError, parseDataUri, postJson, requireEnv } from "./http";

const BASE_URL = "https://generativelanguage.googleapis.com/v1beta/models";

interface GeminiResponse {
  candidates?: Array<{
    content?: { parts?: Array<{ inlineData?: { mimeType?: string; data?: string } }> };
  }>;
}

async function runOne(modelId: string, input: Record<string, unknown>): Promise<string> {
  const prompt = String(input.prompt ?? "");
  const images = Array.isArray(input.images) ? (input.images as string[]) : [];

  const parts: unknown[] = [{ text: prompt }];
  for (const uri of images) {
    const parsed = parseDataUri(uri);
    if (!parsed) throw new ProviderError("google: http image inputs are not supported, data URI expected");
    parts.push({ inline_data: { mime_type: parsed.mime, data: parsed.data } });
  }

  const data = (await postJson(
    `${BASE_URL}/${modelId}:generateContent`,
    { "x-goog-api-key": requireEnv("GOOGLE_API_KEY") },
    { contents: [{ parts }], generationConfig: { responseModalities: ["IMAGE"] } }
  )) as GeminiResponse;

  for (const candidate of data.candidates ?? []) {
    for (const part of candidate.content?.parts ?? []) {
      if (part.inlineData?.data) {
        return `data:${part.inlineData.mimeType ?? "image/png"};base64,${part.inlineData.data}`;
      }
    }
  }
  throw new ProviderError("google: no image in generateContent response");
}

export const googleAdapter: ProviderAdapter = {
  name: "google",
  async run(modelId, input) {
    const count = Math.max(1, Number(input.quantity) || 1);
    const urls = await Promise.all(Array.from({ length: count }, () => runOne(modelId, input)));
    return { images: urls.map((url) => ({ url })) };
  },
};

File: worker/providers/google.py
"""Adaptateur Google Gemini — API officielle (port de google.ts).

Modèles servis : gemini-2.5-flash-image ("Nano Banana"),
gemini-3-pro-image ("Nano Banana Pro") — génération image à partir de
 texte seul et/ou d'images de référence.

Schéma : POST /v1beta/models/{model}:generateContent (header
`x-goog-api-key`) ; le prompt + les images de référence partent en parts
inline_data ; la réponse rend des parts inlineData (base64), converties
ici en data URIs (les workflows les stockent ensuite sur disque).
Une image par requête : la quantité est servie par requêtes parallèles.

Env : GOOGLE_API_KEY (https://aistudio.google.com/apikey) — lue à l'appel,
jamais à l'import (échec vite -> fallback).
"""
from concurrent.futures import ThreadPoolExecutor

from providers.http_helpers import (
    ProviderError,
    parse_data_uri,
    post_json,
    require_env,
)

BASE_URL = "https://generativelanguage.googleapis.com/v1beta/models"


def _run_one(model_id: str, input_: dict) -> str:
    images = input_.get("images")
    if not isinstance(images, list):
        images = []

    parts = [{"text": str(input_.get("prompt") or "")}]
    for uri in images:
        parsed = parse_data_uri(uri)
        if not parsed:
            raise ProviderError("google: http image inputs are not supported, data URI expected")
        mime, data = parsed
        parts.append({"inline_data": {"mime_type": mime, "data": data}})

    response = post_json(
        f"{BASE_URL}/{model_id}:generateContent",
        {"x-goog-api-key": require_env("GOOGLE_API_KEY")},
        {"contents": [{"parts": parts}], "generationConfig": {"responseModalities": ["IMAGE"]}},
    )
    for candidate in response.get("candidates") or []:
        for part in (candidate.get("content") or {}).get("parts") or []:
            inline = part.get("inlineData") or {}
            if inline.get("data"):
                return f"data:{inline.get('mimeType') or 'image/png'};base64,{inline['data']}"
    raise ProviderError("google: no image in generateContent response")


def generate(model_id: str, input_: dict, timeout_ms: int) -> dict:
    """Contrat provider : {"images": [{"url": ...}]}. Appel synchrone (pas
    de polling) : timeout_ms est inutilisé, comme dans le port TS."""
    try:
        count = max(1, int(input_.get("quantity") or 1))
    except (TypeError, ValueError):
        count = 1
    with ThreadPoolExecutor(max_workers=count) as pool:
        urls = list(pool.map(lambda _: _run_one(model_id, input_), range(count)))
    return {"images": [{"url": url} for url in urls]}
